### PR TITLE
Notify request hooks for cors validation failed

### PR DESCRIPTION
**Pull request by miherlosev**
@LavrovArtem 

Changes:
* add `RequestFilterRule.toString` method
* notify request hooks about cors validation failed requests

### CODE REVIEW

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

File: src/request-pipeline/index.js [hunk 1]
```diff
@@ -63,7 +62,14 @@ const stages = {
     2: function checkSameOriginPolicyCompliance (ctx, next) {
         ctx.buildContentInfo();
 
-        if ((ctx.isXhr || ctx.isFetch) && !ctx.contentInfo.isNotModified && !checkSameOriginPolicy(ctx)) {
+        if (!ctx.isKeepSameOriginPolicy()) {
+            ctx.requestFilterRules.forEach(rule => {
+                const configureResponseEvent = new ConfigureResponseEvent(rule, ConfigureResponseEventOptions.DEFAULT);
+
+                ctx.session.callRequestEventCallback(REQUEST_EVENT_NAMES.onConfigureResponse, rule, configureResponseEvent);
+
```
Comment (churkin) [inline src/request-pipeline/index.js]: Extra "new line"

File: src/request-pipeline/index.js [hunk 2]
```diff
@@ -63,7 +62,14 @@ const stages = {
...
+                callOnResponseEventCallbackForFailedSameOriginCheck(ctx, rule, configureResponseEvent);
+            });
             ctx.closeWithError(SAME_ORIGIN_CHECK_FAILED_STATUS_CODE);
             return;
```
Comment (churkin) [inline src/request-pipeline/index.js]: missed "new line"

File: src/request-pipeline/context.js [hunk 3]
```diff
@@ -262,4 +263,13 @@ export default class RequestPipelineContext {
...
         });
     }
+
+    isKeepSameOriginPolicy () {
+        const isNotModifiedAjaxRequest = (this.isXhr || this.isFetch) && !this.contentInfo.isNotModified;
+
+        if (isNotModifiedAjaxRequest)
+            return checkSameOriginPolicy(this);
+
+        return true;
```
Comment (churkin) [inline src/request-pipeline/context.js]: I like it more like this:
```javascript
return isNotModifiedAjaxRequest ? checkSameOriginPolicy(this) : true;
```

File: src/request-pipeline/index.js [hunk 4]
```diff
@@ -268,6 +274,17 @@ function callOnResponseEventCallbackWithCollectedBody (ctx, rule, configureOpts)
     ctx.destRes.pipe(destResBodyCollectorStream);
 }
 
+function callOnResponseEventCallbackForFailedSameOriginCheck (ctx, rule, configureOpts) {
+    const responseInfo  = requestEventInfo.createResponseInfo(ctx);
```
Comment (churkin) [inline src/request-pipeline/index.js]: double space

File: src/request-pipeline/request-hooks/request-filter-rule.js [hunk 5]
```diff
@@ -89,4 +113,11 @@ export default class RequestFilterRule {
     static get ANY () {
         return new RequestFilterRule(MATCH_ANY_REQUEST_REG_EX);
     }
+
+    toString () {
+        if (typeof this.options === 'function')
+            return this._stringifyFunctionOptions();
+
+        return this._stringifyObjectOptions();
```
Comment (churkin) [inline src/request-pipeline/request-hooks/request-filter-rule.js]:
```javascript
const isFunctionOptions = typeof this.options === 'function';

return isFunctionOptions ? this._stringifyFunctionOptions() : this._stringifyObjectOptions();
```

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
The tests fell due to Saucelabs problem. It looks like the Saucelabs virtual machine with IE11 does not have installed Java (see https://www.screencast.com/t/qgbfrly3HGAo)

**Pull request merged**